zjit/src/backend/lir.rs

```diff
@@ -1841,21 +1865,9 @@ impl fmt::Display for Assembler {
                         }
                     } else if let Insn::ParallelMov { moves } = insn {
                         // Print operands with a special syntax for ParallelMov
-                        let mut moves_iter = moves.iter();
-                        if let Some((first_dst, first_src)) = moves_iter.next() {
-                            write!(f, " {first_dst} <- {first_src}")?;
-                        }
-                        for (dst, src) in moves_iter {
-                            write!(f, ", {dst} <- {src}")?;
-                        }
-                    } else {
-                        let mut opnd_iter = insn.opnd_iter();
-                        if let Some(first_opnd) = opnd_iter.next() {
-                            write!(f, " {first_opnd}")?;
-                        }
-                        for opnd in opnd_iter {
-                            write!(f, ", {opnd}")?;
-                        }
+                        write!(f, " {}", moves.iter().map(|(dst, src)| format!("{dst} <- {src}")).collect::<Vec<_>>().join(", "))?;
+                    } else if insn.opnd_iter().count() > 0 {
+                        write!(f, " {}", insn.opnd_iter().map(to_s).collect::<Vec<_>>().join(", "))?;
```

Comment:
This is the `write!(f, "{}", format!())` pattern clippy loves to complaint about because of the extra heap allocation through `String`. Without [`Iterator::intersperse`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.intersperse) I guess the next best thing is `insn.opnd_iter().fold("", |sep, item| { write!(f, "{sep}{item}"); ", "})` But it's a mouthful for code that doesn't need to go super fast.

Up to you which one you like.

Reply:
hah just for completeness it's possible to propagate the error out with try_fold:

```
.try_fold("", |sep, operand| write!(s, "{sep}{operand}").and(Ok(",")))?;
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=f7693eda926c0d4d09549ca88e630f2e

Reply:
nice type gymnastics :) applied the `try_fold` version at https://github.com/ruby/ruby/pull/14992/commits/d377a5cf43667c6fddb94e146736ab451bc6d31b.